src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/StreamExtensions.cs

```diff
@@ -0,0 +1,123 @@
+    {
+        var action = stream.ReadByte();
+        return action switch
+        {
+            0 => ProjectInfoAction.Update,
+            1 => ProjectInfoAction.Remove,
+            _ => throw Assumes.NotReachable()
+        };
+    }
+
+    public static void WriteProjectInfoAction(this Stream stream, ProjectInfoAction projectInfoAction)
+    {
+        stream.WriteByte(projectInfoAction switch
+        {
+            ProjectInfoAction.Update => 0,
+            ProjectInfoAction.Remove => 1,
+            _ => throw Assumes.NotReachable()
+        });
+    }
+
+    public static Task WriteProjectInfoRemovalAsync(this Stream stream, string intermediateOutputPath, CancellationToken cancellationToken)
+    {
+        WriteProjectInfoAction(stream, ProjectInfoAction.Remove);
+        return stream.WriteStringAsync(intermediateOutputPath, cancellationToken: cancellationToken);
+    }
+
+    public static Task<string> ReadProjectInfoRemovalAsync(this Stream stream, CancellationToken cancellationToken)
+    {
+        return stream.ReadStringAsync(cancellationToken: cancellationToken);
+    }
+
+    public static async Task WriteProjectInfoAsync(this Stream stream, RazorProjectInfo projectInfo, CancellationToken cancellationToken)
+    {
+        WriteProjectInfoAction(stream, ProjectInfoAction.Update);
+
+        var bytes = projectInfo.Serialize();
+        using var _ = ArrayPool<byte>.Shared.GetPooledArray(4, out var sizeBytes);
+
+#if NET
+        BitConverter.TryWriteBytes(sizeBytes, bytes.Length);
+#else
+        // Pulled from https://github.com/dotnet/runtime/blob/4b9a1b2d956f4a10a28b8f5f3f725e76eb6fb826/src/libraries/System.Private.CoreLib/src/System/BitConverter.cs#L158C13-L158C87
+        Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(sizeBytes.AsSpan()), bytes.Length);
+#endif
+
+        await stream.WriteAsync(sizeBytes, 0, sizeBytes.Length, cancellationToken).ConfigureAwait(false);
+        await stream.WriteAsync(bytes, 0, bytes.Length, cancellationToken).ConfigureAwait(false);
+    }
+
+    public static async Task<RazorProjectInfo?> ReadProjectInfoAsync(this Stream stream, CancellationToken cancellationToken)
+    {
+        using var _ = ArrayPool<byte>.Shared.GetPooledArray(4, out var sizeBytes);
+        await stream.ReadAsync(sizeBytes, 0, sizeBytes.Length, cancellationToken).ConfigureAwait(false);
+
+        var sizeToRead = BitConverter.ToInt32(sizeBytes, 0);
+
+        using var _1 = ArrayPool<byte>.Shared.GetPooledArray(sizeToRead, out var projectInfoBytes);
+        await stream.ReadAsync(projectInfoBytes, 0, projectInfoBytes.Length, cancellationToken).ConfigureAwait(false);
+        return RazorProjectInfo.DeserializeFrom(projectInfoBytes.AsMemory());
+    }
+
+    private static void WriteSize(Stream stream, int length)
+    {
+        var bytes = GetSizeBytes(length);
+        stream.Write(bytes, 0, bytes.Length);
+    }
+
+    private static int ReadSize(Stream stream)
```

Comment:
Consider `#if NET` this code and use a stack alloc'd span to avoid unnecessary `byte[]` allocation.

Reply:
Done, and used array pool for other case